### PR TITLE
back out locking added in c78ada09f230a3c66cd934860700f93ff31a93eb

### PR DESCRIPTION
**Pull request by algobolson**
This seemed to be needed at the time, but now seems to be causing a major slowdown and removing it doesn't break any tests.

### CODE REVIEW

File: util/execpool/backlog.go [hunk 1]
```diff
@@ -120,18 +102,6 @@ func (b *backlog) Enqueue(enqueueCtx context.Context, t ExecFunc, arg interface{
...
-			return enqueueCtx.Err()
-		case <-b.ctx.Done():
-			return b.ctx.Err()
-		default:
-			return nil
-		}
-	}
 	select {
 	case b.buffer <- backlogItemTask{
 		enqueuedTask: enqueuedTask{
```
Comment (justicz) [inline util/execpool/backlog.go]: github won't let me comment on lines not in this PR, but it looks like this doesn't totally undo the changes done to this file in #700 b/c of the `case <-b.ctx.Done()` case

Comment (justicz) [reply util/execpool/backlog.go]: (in both `Enqueue` and `EnqueueBacklog`)

Comment (algobolson) [reply util/execpool/backlog.go]: This may actually be the minimum change needed to fix things. I think without this Enqueue() and EnqueueBacklog() will panic if Shutdown() has been called because they try to send on a closed b.buffer . The internal Context b.ctx is the only thing that can signal that Enqueue shuoldn't actually try to append to b.buffer

**Comment by algobolson** [comment]
@tsachiherman this fixes the performance regression introduced at #700 ; I have measured it so.

File: util/execpool/backlog.go (hunk 1 above)
Comment (derbear) [reply util/execpool/backlog.go]: Is there a race between context cancellation and closing the buffer?

**Comment by derbear** [comment]
I'm not sure how easy this is, but the backlog struct seems somewhat messy, and it's hard to show that it's correct. Maybe it's worth simplifying this part of the code.

**Comment by algobolson** [comment]
the net diff for backlog.go from before #700 to this is
```
$ git diff 4f4a372f4c204d89457be0ea3a558e77f9f6726b..dd6ee4cf4d35aec8b28cd42cf1472f73cb58e37e .
diff --git a/util/execpool/backlog.go b/util/execpool/backlog.go
index bdce408a..62899fcb 100644
--- a/util/execpool/backlog.go
+++ b/util/execpool/backlog.go
@@ -95,6 +95,8 @@ func (b *backlog) Enqueue(enqueueCtx context.Context, t ExecFunc, arg interface{
                return nil
        case <-enqueueCtx.Done():
                return enqueueCtx.Err()
+       case <-b.ctx.Done():
+               return b.ctx.Err()
        }
 }
 
@@ -112,6 +114,8 @@ func (b *backlog) EnqueueBacklog(enqueueCtx context.Context, t ExecFunc, arg int
                return nil
        case <-enqueueCtx.Done():
                return enqueueCtx.Err()
+       case <-b.ctx.Done():
+               return b.ctx.Err()
        }
 }
```

File: util/execpool/backlog.go [hunk 2]
```diff
@@ -82,25 +78,11 @@ func (b *backlog) GetParallelism() int {
 
 // IsFull test to see if the input buffer is full.
 func (b *backlog) IsFull() bool {
```
Comment (derbear) [inline util/execpool/backlog.go]: Is this method actually called anywhere?

**Comment by tsachiherman** [comment]
Not anymore. I think we should remove it too ( as it seems to require the
lock which we just removed.. )

>

**Comment by tsachiherman** [comment]
I think that the following line is why you're seeing this issue:
```txvalidator.txgroups = make(chan []transactions.SignedTxnWithAD, 10)```

if you were to resize this channel to the size of the groups, it would not make the blocking on the backlog, and could restore the usage profile on the execution pool.

Also, I'm not sure if the #700 change correctly drops expired transactions ( i.e. pre-validation, naturally ).

**Comment by tsachiherman** [comment]
btw - don't take me wrong : This PR could still be making a great performance booster

**Comment by algobolson** [comment]
I believe this change is strictly better than both pre- #700 and better than current master

**Comment by algobolson** [comment]
Further experimentation shows this is still probably only to the level of "probably won't crash" and "slightly less likely to crash than 2 weeks ago"

**Comment by derbear** [comment]
As discussed, though it triggers a panic, the execution race happens only on `goal node stop` and so it's probably not critical to remove it. I'd still like to fix the race condition at some point.
